Remove song import global automation

Veratil opened the pull request:
Completes step 2 from #403 

I've also removed the global automation `<track>` node from the templates. I noticed in `default.mpt` that there were `<object>` nodes, but no related `id` in any other nodes like there would normally be if there was a pattern connected. I believe this might be a bug in the automation editor.

PhysSong commented:
I think `DataFile` is a better place to handle the upgrade stuff rather than `Song`.

Veratil commented:
> I think `DataFile` is a better place to handle the upgrade stuff rather than `Song`.

I tried that initially, but it didn't do anything. I'll try playing with it some more.

Veratil commented:
@PhysSong okay it took some playing around with finding nodes, but I finally managed to get it working in DataFile.

PhysSong commented:
@Veratil Could you look into factory project files to see if any of them has global automation tracks?

Veratil commented:
> @Veratil Could you look into factory project files to see if any of them has global automation tracks?

@PhysSong I believe that's already taken care of in the second commit, or are there other ones that I missed?

PhysSong commented:
It will be okay if you've checked all demo projects and confirmed that there are no global automation tracks that you need to remove.

Veratil commented:
> It will be okay if you've checked all demo projects and confirmed that there are no global automation tracks that you need to remove.

I've gone through and found that some of the demos are still adding tracks that aren't linked to anything. I've updated the check to look for more than one `<time>` node now which has prevented adding extraneous automation tracks which aren't linked to anything.

There is one demo that has an unlinked automation track which has mulitple time nodes.

@PhysSong Would you like me to go through and update all the demo files in this patch to the "newest" save file version then?

PhysSong commented:
> @PhysSong Would you like me to go through and update all the demo files in this patch to the "newest" save file version then?

If you want to do so, I'd recommend not touching files without global automations. I'm also okay if you do that in a separate pull request.

Veratil commented:
> 
> 
> > @PhysSong Would you like me to go through and update all the demo files in this patch to the "newest" save file version then?
> 
> If you want to do so, I'd recommend not touching files without global automations. I'm also okay if you do that in a separate pull request.

Is there an easy way to read the `mmpz` files? That would be the easiest to see if there is any automation patterns that should be imported.

PhysSong commented:
> Is there an easy way to read the `mmpz` files?

`lmms -d input.mmpz` prints the decompressed version of `input.mmpz`.

Veratil commented:
> 
> 
> > Is there an easy way to read the `mmpz` files?
> 
> `lmms -d input.mmpz` prints the decompressed version of `input.mmpz`.

Would it be okay to run `lmms upgrade [in] [out]` on all the files? 😄

PhysSong commented:
That makes sense, but Qt will mix the order of attributes in XML tags. You can write a simple program to compress/decompress `.mmpz` files with `QByteArray::qCompress()` and `QByteArray::qUncompress()`.

PhysSong commented:
The result of `qCompress()` consists of the size of the uncompressed data(in 4byte big-endian integer) and the Zlib-compressed data. It means you can use some other software to (de)compress the project files.

JohannesLorenz commented:
@Veratil Are you going to finish this PR?

Spekular commented:
Seconded. It would be superb if we could get this and step 3 (remove from the GUI) done for 1.3, and step 3 should be a quick and safe fix.

Veratil commented:
I need to go through all the demo files still.

PhysSong commented:
@Veratil Bump.

Veratil commented:
1. I rebased on master 😁 
1. I updated the update function to the new format.
2. I removed the global automation save from Song's `saveSettings`, no more hidden automations will be saved after this at all.
3. I went through each demo file and actually loaded them in lmms. Any that had an unused hidden automation tracks that actually created visible automation tracks I removed and saved over. Anything else I didn't touch. If they do have the block, it's either ignored or will get removed on future updates where we update the demos for whatever reason.

PhysSong commented:
Regarding global automation clips with only one point: In many cases, they are just holding the model's value at time 0, but it's not guaranteed that they always do.

PhysSong commented:
The rest looks good to me.

Veratil commented:
> 
> 
> Regarding global automation clips with only one point: In many cases, they are just holding the model's value at time 0, but it's not guaranteed that they always do.

That's where thing are weird. Every global automation track that had one node was either a leftover that wasn't removed (file bloat because there wasn't a way to delete the global track), or didn't actually affect the model's value anymore (that value was saved as an attribute instead or didn't actually connect anymore).

I can make the change to load it regardless, and I'll go through all the projects again to see if anything else pops up.

PhysSong commented:
I looked into this PR again. This PR will change some behaviors:
- Global automation patterns have lower priority than ordinary ones, but this PR will reverse it due to the order of XML nodes
- Global automation patterns with single point which "locks" the value of connected models will be removed
- Global automation patterns created by user will never be changed

The first one sounds undesirable, but the rest don't sound too bad once #5230 is merged and https://github.com/LMMS/lmms/pull/5223#issuecomment-759225962 is addressed.

PhysSong commented on src/core/DataFile.cpp:
```diff
@@ -1385,6 +1387,44 @@ void DataFile::upgrade_noHiddenClipNames()
+		&& gaTrack.attribute("type").toInt() == Track::HiddenAutomationTrack)
+	{
+		// global automationpatterns
+		QDomNodeList aps = gaTrack.elementsByTagName("automationpattern");
```
This should be changed to `"automationclip"` now.

Veratil replied:
Done!

Veratil commented:
Rebased on master and (hopefully) fixed conflicts (correctly). Tested build and it seemed to be working.
I'll go back through all the demos to double check everything, those are the biggest changes for removing the hidden automation tracks.

Veratil commented:
Looks like everything is still good in the demos. 👍

PhysSong commented:
> Global automation patterns have lower priority than ordinary ones, but this PR will reverse it due to the order of XML nodes

I think this is now the only one which probably needs to be addressed. @Veratil any opinions?

Veratil commented:
> > Global automation patterns have lower priority than ordinary ones, but this PR will reverse it due to the order of XML nodes
> 
> I think this is now the only one which probably needs to be addressed. @Veratil any opinions?

I'm honestly not sure what you mean by priority here. If two automation clips are connected to the same object the global wins?

PhysSong commented:
> If two automation clips are connected to the same object the global wins?

Yes, I mean that case.
https://github.com/LMMS/lmms/blob/1f30ffc5e4dddbd6ffb843bd8c56a8b4de859675/src/core/Song.cpp#L841-L844
Here you can see the global automation track is the first element before removal. You can also see that `TrackContainer::automatedValuesFromTracks` uses the value from the last track connected to an object.

Veratil commented:
Okay I'll try to see what can be done for this to keep them in order. It never occurred to me that you could connect an object to multiple automations.

Rossmaxx commented:
Update?

PhysSong commented:
> Update?

It's ready for review now.

Rossmaxx commented:
Is the track order thing fixed now?

Rossmaxx commented on data/projects/templates/AcousticDrumset.mpt:
```diff
@@ -114,8 +114,8 @@
       <automationpattern prog="0" mute="0" name="Master volume" pos="0" tens="1" len="192"/>
       <automationpattern prog="0" mute="0" name="Master pitch" pos="0" tens="1" len="192"/>
     </track>
-    <mixer visible="1" width="561" height="332" x="10" y="314" maximized="0" minimized="0">
```
Why is mixer renamed to fxmixer again? Is this some regression?

Rossmaxx replied:
You gotta look for this at mmpz files too.

Rossmaxx commented on data/projects/demos/StrictProduction-DearJonDoe.mmp:
```diff
@@ -1,643 +1,670 @@
 <?xml version="1.0"?>
 <!DOCTYPE lmms-project>
-<lmms-project creator="LMMS" version="1.0" type="song" creatorversion="1.2.0">
```
Is this force resave necessary? If so, what are these new tags?

Rossmaxx commented on data/projects/demos/StrictProduction-DearJonDoe.mmp:
```diff
@@ -1,643 +1,670 @@
               </pattern>
             </track>
           </trackcontainer>
         </bbtrack>
```
I thought we renamed it to pattern

Rossmaxx replied:
The line below

Rossmaxx commented on data/projects/templates/CR8000.mpt:
```diff
@@ -232,8 +232,8 @@
     </track>
-    <mixer visible="1" width="561" height="332" x="10" y="314" maximized="0" minimized="0">
-      <mixerchannel num="0" muted="0" volume="1" name="Master" soloed="0">
+    <fxmixer visible="1" width="561" height="332" x="10" y="314" maximized="0" minimized="0">
```
All these other presets seems to have this change.

Rossmaxx reviewed (commented):
Got a few concerns from the diff.

PhysSong commented:
@Rossmaxx Factory projects probably requires re-saving again, I think

Rossmaxx commented:
Yeah, or unstaging them for now.

PhysSong commented:
> Yeah, or unstaging them for now.

I think we can defer re-saving to right before releasing 1.3 or any other sane point. Unstaging for now.

I also manually removed related parts in factory templates and tutorials.

Rossmaxx reviewed (approved):
Everything looks good for now.

Rossmaxx commented:
Removing global automation shouldn't have waited this long.

michaelgregorius commented on src/core/Song.cpp:
```diff
@@ -1239,7 +1233,6 @@ bool Song::saveProjectFile(const QString & filename, bool withResources)
 
 	saveState( dataFile, dataFile.content() );
 
-	m_globalAutomationTrack->saveState( dataFile, dataFile.content() );
```
Shouldn't the member `m_globalAutomationTrack` be fully removed as well?

PhysSong replied:
Yes, but in another PR.

PhysSong replied:
To elaborate, That is the step 4 in #403, which should be done after #5230(step 3).

michaelgregorius replied:
Ok, thanks for the heads up! :+1:

michaelgregorius commented on src/core/DataFile.cpp:
```diff
@@ -1638,6 +1639,51 @@ void DataFile::upgrade_1_3_0()
+		&& gaTrack.attribute("type").toInt() == static_cast<int>(Track::Type::HiddenAutomation))
+	{
+		// global automationclips
+		QDomNodeList aps = gaTrack.elementsByTagName("automationclip");
```
IMO the code would become better readable if that was called `automationClips` instead of `aps`. What does "aps" stand for?

PhysSong replied:
I think `ap` stands for `AutomationPattern` which is `AutomationClip` now. I agree on renaming.

PhysSong replied:
Renamed variables.

michaelgregorius commented on src/core/DataFile.cpp:
```diff
@@ -1638,6 +1639,51 @@ void DataFile::upgrade_1_3_0()
+		QList<QDomNode> tracksToInsert;
+		for (int i = 0; i < aps.length(); ++i)
+		{
+			QDomElement ap = aps.item(i).toElement();
```
Similar as above. Would be better to understand if `ap` was called `automationClip`.

michaelgregorius commented on src/core/DataFile.cpp:
```diff
@@ -1638,6 +1639,51 @@ void DataFile::upgrade_1_3_0()
+			if (ap.elementsByTagName("time").length() > 1)
+			{
+				QDomElement aptrack = createElement("track");
+				aptrack.setAttribute("muted", QString::number(false));
```
In general it's a bit unfortunate that the upgrade code must know the implementation details of how other entities represent their data in the XML file. However, that's a design problem and not something for this PR.

Rossmaxx commented:
Shiit, acted too soon. A test from @bratpeki would have helped. But now, you can test on master too.

bratpeki commented:
It's cool, I'll take a look at master after the weekend! Can you give me a TLDR of all changes made here?

sakertooth commented:
Should the global automation test in `AutomationTrackTest` been removed?

Rossmaxx commented:
I believe we should, but I'll wait for @PhysSong 's opinion too

PhysSong commented:
Maybe we could do as part of step 4 of #403.
